src/shared/utils/s3.ts

```diff
@@ -52,12 +62,80 @@ export async function fetchS3(bucket: string, key: string): Promise<string> {
   });
 
   const response = await send(getCommand);
-  const data = await response.Body?.transformToString();
-  if (data === undefined) throw new Error(`No data found in ${bucket}/${key}`);
+  const responseBody = response.Body;
+
+  if (responseBody === undefined)
+    throw new Error(`No data found in ${bucket}/${key}`);
+
+  const metadata = response.Metadata ?? {};
+
+  if (isEncryptedS3ObjectMetadata(metadata)) {
+    const responseBodyBytes = await responseBody.transformToByteArray();
+    return await getDecrypted(responseBodyBytes, metadata);
+  }
 
-  return data;
+  return await responseBody.transformToString();
 }
 
+const getDecrypted = async (
+  encryptedBytes: Uint8Array,
+  metadata: EncryptedS3ObjectMetadata
+): Promise<string> => {
+  const {
+    "x-amz-cek-alg": encryptionAlgorithmName,
+    "x-amz-iv": initialisationVectorBase64Text,
+    "x-amz-key-v2": encryptedKey,
+    "x-amz-matdesc": keyContextText,
+    "x-amz-tag-len": authTagBitCountText,
+  } = metadata;
+
+  if (encryptionAlgorithmName !== "AES/GCM/NoPadding")
+    throw Error(`Unsupported encryption algorithm: ${encryptionAlgorithmName}`);
+
+  const encryptedKeyBuffer = Buffer.from(encryptedKey, "base64");
+
+  const keyContext = JSON.parse(keyContextText);
+  if (typeof keyContext !== "object" || keyContext === null)
+    throw Error("Invalid key context");
+
+  const key = await decryptKms(encryptedKeyBuffer, keyContext);
+
+  const authTagLength = parseInt(authTagBitCountText, 10) / 8;
```

Comment:
More a question than a comment...  This looks like it can easily take on a non-integer value, but you then use it in slice().  From what I've googled, slice will internally coerce it to an integer, but are you confident this will always work as expected?